pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionPool.java

```diff
@@ -357,5 +418,46 @@ int getPoolSize() {
 
     private static final Logger log = LoggerFactory.getLogger(ConnectionPool.class);
 
+    public void doMarkAndReleaseUselessConnections(){
+        if (!autoReleaseIdleConnectionsEnabled){
+            return;
+        }
+        List<Runnable> releaseIdleConnectionTaskList = new ArrayList<>();
+        for (Map.Entry<InetSocketAddress, ConcurrentMap<Integer, CompletableFuture<ClientCnx>>> entry :
+                pool.entrySet()){
+            ConcurrentMap<Integer, CompletableFuture<ClientCnx>> innerPool = entry.getValue();
+            for (Map.Entry<Integer, CompletableFuture<ClientCnx>> entry0 : innerPool.entrySet()){
+                CompletableFuture<ClientCnx> future = entry0.getValue();
+                // Ensure connection has been connected.
+                if (!future.isDone()){
+                    continue;
+                }
+                if (future.isCompletedExceptionally()){
+                    continue;
+                }
+                try {
+                    final ClientCnx clientCnx = future.get();
```

Comment:
We can use `join` to avoid catching useless exceptions.

Reply:
Good idea. Already fixed.